File: csharp/dsf/dsf/src/pool/ClientCreator.cs
using System;

namespace com.baidu.dsf.pool
{
    using Thrift.Protocol;
    using Thrift.Transport;
    using com.baidu.dsf.register;
    using com.baidu.dsf.thrift;
    using Thrift;

    /// 
    /// <summary>
    /// @author arksea
    /// </summary>
    public class ClientCreator<I>
    {
        private readonly MsgServiceDefine serviceInfo;

        public ClientCreator(MsgServiceDefine serviceInfo)
        {
            this.serviceInfo = serviceInfo;
        }

        public virtual TServiceClient create(ServiceInstance inst, int timeout)
        {
            TSocket sock = null;
            MsgSvcAddr svcAddr = inst.Addr;
            try
            {
                sock = new TSocket(svcAddr.host, svcAddr.port, timeout);
                sock.Open();
                string protocolStr = serviceInfo.getProperty("protocol");
                if (protocolStr == null)
                {
                    throw new ApplicationException("service not defined protocol: " + serviceInfo.regname);
                }
                TProtocol protocol = ProtocolFactory.create(protocolStr, sock);
                string className = serviceInfo.name + "+Client";
                System.Reflection.Assembly assembly =System.Reflection.Assembly.GetAssembly(typeof(I));
                Type clientClass = assembly.GetType(className);
                var client = (TServiceClient)Activator.CreateInstance(clientClass, protocol);
                inst.onConnectSucceed();
                return client;
            }
            catch (Exception ex)
            {
                if (sock != null)
                {
                    sock.Close();
                }
                inst.onConnectFailed(ex);
                string err = "Create client failed：" + serviceInfo.regname + "@" + svcAddr.host + ":" + svcAddr.port;
                throw new ClientSourceException(err, serviceInfo.regname, svcAddr, ex);
            }
        }
    }

}

File: java/dsframework/src/main/java/dsf/pool/ClientCreator.java
package dsf.pool;

import java.lang.reflect.Constructor;

import org.apache.thrift.TServiceClient;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import dsf.register.MsgServiceDefine;
import dsf.register.MsgSvcAddr;
import dsf.register.ServiceInstance;
import dsf.thrift.ProtocolFactoryUtil;

/**
 *
 * @author arksea
 */
public class ClientCreator implements IClientCreator<TServiceClient> {

    private final static Logger logger = LogManager.getLogger(ClientFactory.class);
    private final MsgServiceDefine serviceInfo;

    public ClientCreator(MsgServiceDefine serviceInfo) {
        this.serviceInfo = serviceInfo;
    }

    @Override
    public TServiceClient createClient(ServiceInstance inst, int timeout) throws ClientSourceException {
        TSocket sock = null;
        MsgSvcAddr svcAddr = inst.getAddr();
        try {
            sock = new TSocket(svcAddr.host, svcAddr.port, timeout);
            sock.open();
            String protocolStr = serviceInfo.getProperty("protocol");
            assert (protocolStr != null);
            TProtocol protocol = ProtocolFactoryUtil.getFactory(protocolStr).getProtocol(sock);
            String className = serviceInfo.name + "$Client";
            Class<?> clientClass = Class.forName(className);
            Constructor con = clientClass.getConstructor(TProtocol.class);
            TServiceClient client = (TServiceClient) con.newInstance(protocol);
            inst.onConnectSucceed();
            return client;
        } catch (Throwable ex) {
            if (sock != null) {
                sock.close();
            }
            inst.onConnectFailed(ex);
            String err = "Create client failed：" + serviceInfo.regname + "@" + svcAddr.host + ":" + svcAddr.port;
            logger.debug(err, ex);
            throw new ClientSourceException(err, serviceInfo.regname, svcAddr, ex);
        }
    }
}
